File: 3Sum/solution.swift

class Solution {
    func threeSum(_ nums: [Int]) -> [[Int]] {
        guard nums.count > 2 else { return [] }
        var res = [[Int]]()
        var found = Set<String>()
        var nums = nums.sorted()
        
        for p1 in 0..<nums.count - 2 {
            var p2 = p1 + 1 
            var p3 = nums.count - 1 
            
            while p2 < p3 {
                var sum = nums[p1] + nums[p2] + nums[p3]
                if sum == 0 {
                    var key = "\(nums[p1]) \(nums[p2]) \(nums[p3])"
                    if !found.contains(key) {
                        found.insert(key)
                        res.append([nums[p1], nums[p2], nums[p3]])
                    }
                
                p2 += 1 
                p3 -= 1
                } else if sum < 0 {
                    p2 += 1 
                } else {
                    p3 -= 1
                }
            }
        }
        
        return res 
    }
}

File: 3Sum/solution.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */

var threeSum = function(nums) {
    var keys = new Set([]);
    var answers = [];

    nums.sort(function(a, b){ 
        return a - b
    })

    for (var n1 = 0; n1 < nums.length - 1; n1++) {
        n2 = n1 + 1; 
        n3 = nums.length - 1
        
        while (n2 < n3) {
            let sum = nums[n1] + nums[n2] + nums[n3];
            if (sum === 0) {
                let key = `${nums[n1]}/${nums[n2]}/${nums[n3]}`
                if (!keys.has(key)) {
                    answers.push([nums[n1], nums[n2], nums[n3]]);
                    keys.add(key);
                }
                n2++; 
                n3--;
            } else if (sum < 0) {
                n2++; 
            } else {
                n3--; 
            }
        }
    }
    
    return answers;
};


